Remove metric processor Subtract function; rename ExportKind to AggregationTemporality

jmacd opened the pull request:
The basic Processor has code to convert cumulative values to delta values on the export path. This was exploratory functionality, and _has been requested a few times_, but is not considered widely useful. This removes it completely.

There are not many known use-cases for this support for it to be included in the basic processor, but for example:

- Metric backend prefers DELTA to GAUGE data for SumObserver or UpDownSumObserver. For example, monitor change in `process.cpu.time` as a DELTA not a CUMULATIVE.
- Metric exporter configures label removal an `UpDownSumObserver` instrument, with say 2 dimensions input by the observer callback and 1 dimension output (using `Histogram` type); then configures exemplar output (using `RawValue` type) with the complete two dimensions; exemplar sampling should be performed using DELTA weights to select the items most responsible for the rise/fall of the sum.

As you can see these are a bit contrived, and I think the exporters that want these things can easily take a DIY approach.

This PR also renames `ExportKind` to `AggregationTemporality`, as this terminology was already flagged as confusing in an open spec PR: https://github.com/open-telemetry/opentelemetry-specification/pull/1159#discussion_r520960248

jmacd commented:
@open-telemetry/specs-metrics-approvers As I am trying to bring the OTel-Go metrics SDK into line with "Everything we will specify", this is something I'd like to remove. (Note: I'm hoping to relieve the @open-telemetry/go-approvers of this responsibility, it's just code deletion.)

krnowak commented on sdk/export/metric/metric.go:
```diff
@@ -418,12 +407,6 @@ func CumulativeExportKindSelector() ExportKindSelector {
 	return ConstantExportKindSelector(CumulativeExportKind)
 }
 
-// DeltaExportKindSelector returns an ExportKindSelector that
```
Why dropping it? The `DeltaExportKind` still exists, so there could be a 3rd party Processor that actually supports converting precomputed sums to delta. Dropping `DeltaExportKindSelector` would force the existing users to change that to `ConstantExportKindSelector(DeltaExportKind)`, no?

jmacd replied:
Ah, that's a terminology problem that we've marked for fixing in my derelict spec PR, see here: https://github.com/open-telemetry/opentelemetry-specification/pull/1159#discussion_r520960248. `ExportKind` should become `AggregationTemporality` but I would like to do that in a follow-on PR since it touches all the same files.

jmacd replied:
I realized that the confusion I created has a lot to do with the misleading name `ExportKind`, so I added it in this PR. The new diff is pretty large, but it's almost entirely simple text substitution see 74c45ea.

evantorrie commented:
What was an example use-case for this functionality?  Something like an absolute system counter (# of network packets sent) that the backend system wants exposed as a delta?

I presume this does *not* affect any synchronous instruments, which will still support being exported as a delta even if they are an absolute count?

jmacd commented:
Yes @evantorrie something like that. I updated the PR description to clarify some use-cases, but I think the term confusion has a lot to do with "ExportKind", which I've renamed as we have already discussed in an open spec PR (https://github.com/open-telemetry/opentelemetry-specification/pull/1159#discussion_r520960248). 

> The DeltaExportKind still exists, so there could be a 3rd party Processor that actually supports converting precomputed sums to delta.

@krnowak Yes, but I don't really think there are any of these out there. An exporter could use `ConstantExportKindSelector(DeltaExportKind)` and a custom processor could implement it, but it's more complexity than the case processor deserves.

jmacd commented:
@bogdandrutu Please review.

bogdandrutu commented on exporters/otlp/internal/transform/metric.go:
```diff
@@ -398,17 +398,17 @@ func gaugePoint(record export.Record, num number.Number, start, end time.Time) (
 }
 
-func exportKindToTemporality(ek export.ExportKind) metricpb.AggregationTemporality {
+func exportKindToTemporality(ek export.AggregationTemporality) metricpb.AggregationTemporality {
```
This func name needs to be updated.

jmacd replied:
Fixed.

bogdandrutu commented on sdk/metric/processor/basic/basic.go:
```diff
@@ -171,13 +163,17 @@ func (b *Processor) Process(accum export.Accumulation) error {
+				// value and one for the output delta.
+				// This functionality was removed from
+				// the basic processor in PR #1415.
+				return aggregation.ErrNoCumulativeToDelta
```
What do you do with this error in the caller? I think this should be returned to the user when setting the expected temporality instead of framework dealing with this error.

jmacd replied:
The reason this isn't returned earlier is that the decision is passed as a function to allow per-instrument configuration. I guess the functionality here is fairly low-level, and if we get to a higher-level configuration mechanism that layer will have to verify something like this.  There's a `CumulativeAggregationTemporalitySelector` but not a corresponding `DeltaAggregationTemporalitySelector`, so the caller has to go out of their way to configure this condition.

jmacd commented:
PTAL

jmacd commented:
@bogdandrutu please take a look.

MrAlias commented:
Closing as the metrics signal refactor at the specification level might make this work wasted. We can open again if the specification solidifies and it is determined needed.
